### PR TITLE
Refactor ConstraintsLib and InputLib plus added new flags

### PR DESCRIPTION
**Pull request by k06a**
```
·····································|··········|··········|··········|·······
|  fillContractOrderRFQToWithPermit  ·   78422  ·   95522  ·   86972  ·   2  │
·····································|··········|··········|··········|·······
|  fillOrder                         ·   79451  ·  130802  ·  105144  ·  39  │
·····································|··········|··········|··········|·······
|  fillOrderRFQ                      ·   80064  ·  117832  ·   98560  ·   7  │
·····································|··········|··········|··········|·······
|  fillOrderRFQToWithPermit          ·  125855  ·  125979  ·  125917  ·   2  │
·····································|··········|··········|··········|·······
|  fillOrderToWithPermit             ·  131360  ·  131502  ·  131431  ·   2  │
·····································|··········|··········|··········|·······
```
=>
```
·····································|··········|··········|··········|·······
|  fillContractOrderRFQToWithPermit  ·   78274  ·   95374  ·   86824  ·   2  │
·····································|··········|··········|··········|·······
|  fillOrder                         ·   79282  ·  130663  ·  105117  ·  39  │
·····································|··········|··········|··········|·······
|  fillOrderRFQ                      ·   79927  ·  117720  ·   97975  ·  10  │
·····································|··········|··········|··········|·······
|  fillOrderRFQToWithPermit          ·  125750  ·  125874  ·  125812  ·   2  │
·····································|··········|··········|··········|·······
|  fillOrderToWithPermit             ·  131333  ·  131427  ·  131380  ·   2  │
·····································|··········|··········|··········|·······
```

### CODE REVIEW

File: contracts/OrderRFQLib.sol [hunk 1]
```diff
@@ -5,12 +5,32 @@ pragma solidity 0.8.17;
 import "@1inch/solidity-utils/contracts/libraries/ECDSA.sol";
 import "@1inch/solidity-utils/contracts/libraries/AddressLib.sol";
 
+type TraitsRFQ is uint256;
+
+library TraitsRFQLib {
```
Comment (typicalbuster) [inline contracts/OrderRFQLib.sol:10]: Why "Traits" instead of "Details"?

Comment (k06a) [reply contracts/OrderRFQLib.sol:10]: It is kind of restrictions

File: contracts/OrderMixin.sol [hunk 2]
```diff
@@ -119,76 +118,87 @@ abstract contract OrderMixin is IOrderMixin, EIP712, PredicateHelper {
...
-                bytes calldata permitCalldata = permit[20:];
-                IERC20(token).safePermit(permitCalldata);
-                if (_remaining[orderHash] != _ORDER_DOES_NOT_EXIST) revert ReentrancyDetected();
+            if (!input.skipOrderPermit()) {
+                bytes calldata permit = order.permit();
+                if (permit.length >= 20) {
+                    // proceed only if taker is willing to execute permit and its length is enough to store address
+                    address token = address(bytes20(permit));
+                    bytes calldata permitCalldata = permit[20:];
+                    IERC20(token).safePermit(permitCalldata); // TODO: inline both arguments
```
Comment (typicalbuster) [inline contracts/OrderMixin.sol:156]: Why not TODO now?

Comment (k06a) [reply contracts/OrderMixin.sol:156]: Can be refactored simultaneously in multiple places

**Review by typicalbuster — APPROVED** [review]
LGTM

**Pull request merged**